### PR TITLE
Feat/toast

### LINKED ISSUE #162: Toast notifications

### PR DESCRIPTION
**Pull request by JDMathew**
Toast component library setup and demo screen

https://user-images.githubusercontent.com/17250443/187008239-17825689-3d18-442a-8122-42a10fc64758.mp4

resolves #162

### CODE REVIEW

**Comment by keithluchtel** [comment]
I'm noticing a weird animation issue when the toast is animating out. See in the attached video, the icon is switching from red to blue.

https://user-images.githubusercontent.com/475007/187235184-61935069-7050-4413-ad75-6f80e13c4683.mp4

File: libs/component-library/src/lib/toast/components/BaseToast.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      style={s.base}
+      disabled={!onPress}
+      onPress={() => onPress(props)}
+    >
+      {renderIcon?.({ color, ...props }) ?? (
+        <FxBox style={s.iconContainer}>
+          {iconElement ? (
+            React.cloneElement(iconElement, { fill: color })
+          ) : (
+            <FxBox style={s.icon} />
```
[the remark this answers is not in the document]
Comment (JDMathew) [reply libs/component-library/src/lib/toast/components/BaseToast.tsx:48]: It updates the icon with the correct color for the notifications. I.e (success, info, warning...)

[the remark this answers is not in the document]
Comment (keithluchtel) [reply libs/component-library/src/lib/toast/components/BaseToast.tsx:48]: I was more-so referring to line 48: `<FxBox style={s.icon} />` Is this just an empty box?

File: libs/component-library/src/lib/toast/context/ToastProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      ...defaultContext.defaults,
+      ...defaults,
+    },
+    customToasts,
+  };
+
+  return (
+    <ToastContext.Provider value={value} {...rest}>
+      {children}
+      {renderToaster && <Toast />}
```
Comment (keithluchtel) [inline libs/component-library/src/lib/toast/context/ToastProvider.tsx:56]: Do Toasts only render at the top of the screen?

File: libs/component-library/src/lib/toast/components/Toast.tsx [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+  const translationY = useSharedValue(hiddenY);
+
+  const reanimatedStyle = useAnimatedStyle(() => {
+    return {
+      position: 'absolute',
+      alignItems: 'center',
+      justifyContent: 'center',
+      left: 0,
+      top: 0,
+      right: 0,
```
[the remark this answers is not in the document]
Comment (JDMathew) [reply libs/component-library/src/lib/toast/components/Toast.tsx]: Done

File: libs/component-library/src/lib/toast/components/Toast.tsx [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    return toastComponent({
+      ...defaults,
+      ...activeToast,
+      onClose: hideToast,
+      onPress: onPressCallback,
+    });
+  };
+
+  return (
+    <Reanimated.View onLayout={onLayout} style={[reanimatedStyle]}>
```
[the remark this answers is not in the document]
Comment (JDMathew) [reply libs/component-library/src/lib/toast/components/Toast.tsx]: Done

File: libs/component-library/src/lib/toast/context/ToastProvider.tsx (hunk 2 above)
Comment (JDMathew) [reply libs/component-library/src/lib/toast/context/ToastProvider.tsx:56]: Yes, I spent some time trying to add bottom functionality but ended up with buggy behaviour and so for now have kept it at the top

**Comment by JDMathew** [comment]
> I'm noticing a weird animation issue when the toast is animating out. See in the attached video, the icon is switching from red to blue.
> 
>  toast-animation.mp4

Looking into refactoring the order of events with a exit prop to fix this

File: libs/component-library/src/lib/toast/components/Toast.tsx [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+      setInProgress(false);
+      onHide?.(prevToastRef.current as ToastProps);
+      prevToastRef.current = null;
+    };
+
+    const show = async () => {
+      if (inProgress || isVisible) {
+        await hide();
+      }
+      setInProgress(true);
```
[the remark this answers is not in the document]
Comment (JDMathew) [reply libs/component-library/src/lib/toast/components/Toast.tsx:130]: If a toast is added while it's in progress it doesn't run show or hide

File: libs/component-library/src/lib/toast/components/BaseToast.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (JDMathew) [reply libs/component-library/src/lib/toast/components/BaseToast.tsx:49]: It adds a spacer for where the icon would be so you have consistent layout between notifications that have icons and ones that don't

File: libs/component-library/src/lib/toast/components/Toast.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (JDMathew) [reply libs/component-library/src/lib/toast/components/Toast.tsx]: Have removed

File: libs/component-library/src/lib/toast/components/Toast.tsx [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    // no toasts or correct one is already shown
+    if (
+      activeToast === prevToastRef.current ||
+      (!activeToast && !prevToastRef.current)
+    ) {
+      return;
+    }
+
+    const hide = async () => {
+      setInProgress(true);
```
[the remark this answers is not in the document]
Comment (JDMathew) [reply libs/component-library/src/lib/toast/components/Toast.tsx]: have removed

**Comment by JDMathew** [comment]
Updated (with slow animation out)

https://user-images.githubusercontent.com/17250443/187772557-88cb9f2c-2db9-4602-bdfc-3f0d163ed5fa.mp4

**Review by keithluchtel — APPROVED** [review]

File: libs/component-library/src/lib/toast/components/Toast.tsx [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+  const [isVisible, setIsVisible] = useState(false);
+
+  const onHide = defaults.onHide;
+
+  const onShow = useMemo(
+    () => activeToast?.onShow ?? defaults.onShow,
+    [activeToast?.onShow, defaults.onShow]
+  );
+
+  const prevHeightRef = useRef<number | null>();
```
[the remark this answers is not in the document]
Comment (JDMathew) [reply libs/component-library/src/lib/toast/components/Toast.tsx]: outdated

**Pull request merged**